### PR TITLE
[Authz] Support building function privileges in Spark 3.4

### PR DESCRIPTION
**Pull request by packyan**
### _Why are the changes needed?_

Add support for function privileges building in 3.4 for #4167 

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by packyan** [comment]
The reason why the previous pr4167 failed to extract the privilege object of the function on 3.4 is that:

1 spark 3.4 adds catalog to the name of the functions, from `database.function_name` to `catalog.database.function_name`, so `buildFunctionIdentFromQualifiedName` needs to be adapted for spark 3.4.

2. The behavior of the `transformAllExpressions` method of `LogicPlan` has been changed in 3.4. Its subqueries can not be traversed. So I added a branch to the buildFunction method to handle Commands with subqueries.

cc @bowenliang123 , also cc @yaooqinn .

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala [hunk 1]
```diff
@@ -220,15 +220,26 @@ object PrivilegesBuilder {
...
-    plan transformAllExpressions {
-      case hiveFunction: Expression if isKnownFunction(hiveFunction) =>
-        val functionSpec: ScanSpec = getFunctionSpec(hiveFunction)
-        if (functionSpec.functionDescs.exists(!_.functionTypeDesc.get.skip(hiveFunction, spark))) {
-          functionSpec.functions(hiveFunction).foreach(func =>
-            inputObjs += PrivilegeObject(func))
+    plan match {
+      case _: Command =>
+        val classname = plan.getClass.getName
+        if (TABLE_COMMAND_SPECS.contains(classname)) {
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:226]: use `isKnownTableCommand` instead?

Comment (packyan) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:226]: force pushed, please refresh.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala [hunk 2]
```diff
@@ -30,9 +31,12 @@ object FunctionExtractor {
   }
 
   def buildFunctionIdentFromQualifiedName(qualifiedName: String): (String, Option[String]) = {
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala:33]: How about make it return `Function` class directly?

Comment (packyan) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala:33]: > How about make it return `Function` class directly?

reasonable

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala (hunk 1 above)
Comment (packyan) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala]: > use `isKnownTableCommand` instead?

already done.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala [hunk 3]
```diff
@@ -29,13 +30,17 @@ object FunctionExtractor {
...
+    val parts: Array[String] = qualifiedName.split("\\.", partsLen)
+    val (functionName, database) = if (parts.length == 1) {
+      (None, qualifiedName)
+    } else if (parts.length == 3) {
+      (Some(parts.tail.head), parts.last)
     } else {
-      (parts.last, Some(parts.head))
+      (Some(parts.head), parts.last)
     }
+    Function(functionName, database)
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala:43]: Seems reversed.

**Comment by bowenliang123** [comment]
Are you considering extending `Function` to add `catalog` field for Spark 3.4?

**Comment by packyan** [comment]
> 



> Are you considering extending `Function` to add `catalog` field for Spark 3.4?

Add `catalog` field to `Function` is fine, but it seems that there is no place to use the `catalog` field of `Function` currently.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala (hunk 3 above)
Comment (packyan) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala]: > Seems reversed.

I adjust the function field order to kepp consistent with spark.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Function.scala [hunk 4]
```diff
@@ -22,7 +22,8 @@ package org.apache.kyuubi.plugin.spark.authz.serde
  *
  * Represents a function identity
  *
- * @param database
  * @param functionName
+ * @param database
+ * @param catalog
  */
-case class Function(database: Option[String], functionName: String)
+case class Function(functionName: String, database: Option[String], catalog: Option[String])
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/Function.scala:29]: According to other identity types like `Table` and `Database`, it's preferred adding `catalog` in the first place to follow the same pattern in `catalog - db/namespace - object` style.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala [hunk 5]
```diff
@@ -29,13 +30,17 @@ object FunctionExtractor {
     loadExtractorsToMap[FunctionExtractor]
   }
 
-  def buildFunctionIdentFromQualifiedName(qualifiedName: String): (String, Option[String]) = {
-    val parts: Array[String] = qualifiedName.split("\\.", 2)
-    if (parts.length == 1) {
-      (qualifiedName, None)
+  private[authz] def buildFunctionFromQualifiedName(qualifiedName: String): Function = {
+    val partsLen = if (isSparkV34OrGreater) 3 else 2
+    val parts: Array[String] = qualifiedName.split("\\.", partsLen)
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala:35]: What's the reason for setting the limit for splitting?
Even if the version affects the split array length, we could just fetch them by corresponding index.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala [hunk 6]
```diff
@@ -29,13 +30,17 @@ object FunctionExtractor {
...
+    val (functionName, database, catalog) = if (parts.length == 1) {
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/functionExtractors.scala:36]: Better reverting to the correct order for readability as it does for Function construction .

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala [hunk 7]
```diff
@@ -220,15 +220,24 @@ object PrivilegesBuilder {
...
-        if (functionSpec.functionDescs.exists(!_.functionTypeDesc.get.skip(hiveFunction, spark))) {
-          functionSpec.functions(hiveFunction).foreach(func =>
-            inputObjs += PrivilegeObject(func))
+    plan match {
+      case command: Command if isKnownTableCommand(command) =>
+        val spec = getTableCommandSpec(command)
+        spec.queries(plan)
+          .map(plan => buildFunctions(plan, spark)._1)
+          .reduce(_ ++ _)
+          .foreach(functionPriv => inputObjs += functionPriv)
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:229]: Maybe separate the output of `buildFunctions` and the following operations for better understanding. Hence, it implies that only input objects are appended. Output objects could be considered in the future.

**Review by bowenliang123 — APPROVED** [review]
LGTM overall. Several comments for improvement. And please fix the conflicts as well.

**Review by bowenliang123 — APPROVED** [review]

**Comment by bowenliang123** [comment]
Thanks for continuous contributions. Merged to master.

**Comment by pan3793** [comment]
Thanks @packyan and @bowenliang123

**Pull request closed without merging**